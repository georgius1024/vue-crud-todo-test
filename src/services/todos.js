import http from "axios";
const baseUrl = process.env.VUE_APP_ENDPOINT;
const endpoint = `${baseUrl}/todos`;
export function list() {
  return http.get(endpoint).then((response) => {
    return response.data;
  });
}

export function add(todo) {
  return http.post(endpoint, todo).then((response) => {
    return response.data["_id"];
  });
}

export function update(id, todo) {
  return http.put(`${endpoint}/${id}`, todo);
}

export function remove(id) {
  return http.delete(`${endpoint}/${id}`);
}
