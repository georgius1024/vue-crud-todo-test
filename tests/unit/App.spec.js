jest.mock("@/services/todos");
import * as todos from "@/services/todos";
import { shallowMount } from "@vue/test-utils";
import component from "@/App.vue";
import flushPromises from "flush-promises";

describe("app", () => {
  const list = [
    { _id: 101, name: "todo item", done: true },
    { _id: 102, name: "another todo item", done: true },
  ]
  beforeEach(() => {
    todos.list = jest.fn().mockResolvedValue([...list]);
  })
  it("mounts without errors", async () => {
    const wrapper = shallowMount(component);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.html()).toMatchSnapshot();
    await flushPromises();
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
    expect(todos.list).toHaveBeenCalled();
    expect(wrapper.vm.todos).toHaveLength(2);
    expect(wrapper.vm.todos[0]).toHaveProperty("name", "todo item");
  });
  it("adds todo item", async () => {
    const wrapper = shallowMount(component);
    await flushPromises();
    await wrapper.vm.$nextTick();
    wrapper.setData({ text: "new todo item" });
    todos.add = jest.fn().mockResolvedValue(103);
    await wrapper.vm.add();
    await wrapper.vm.$nextTick();
    expect(todos.add).toHaveBeenCalled();
    expect(wrapper.vm.todos).toHaveLength(3);
    expect(wrapper.vm.todos[2]).toHaveProperty("name", "new todo item");
  });
  it("removes todo item", async () => {
    const wrapper = shallowMount(component);
    await flushPromises();
    await wrapper.vm.$nextTick();
    todos.remove = jest.fn().mockResolvedValue();
    await wrapper.vm.remove(wrapper.vm.todos[0]);
    await wrapper.vm.$nextTick();
    expect(todos.remove).toHaveBeenCalled();
    expect(wrapper.vm.todos).toHaveLength(1);
  });
  it("toggles todo item", async () => {
    const wrapper = shallowMount(component);
    await flushPromises();
    await wrapper.vm.$nextTick();
    todos.update = jest.fn().mockResolvedValue();
    await wrapper.vm.toggle(wrapper.vm.todos[0]);
    await wrapper.vm.$nextTick();
    expect(todos.update).toHaveBeenCalled();
    expect(wrapper.vm.todos[0]).toHaveProperty("done", false);
  });
});
