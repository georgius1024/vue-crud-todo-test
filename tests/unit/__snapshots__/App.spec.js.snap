// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app mounts without errors 1`] = `
<section class="">
  <div>Loading...</div>
  <hr><input class="p6" type="text"><button class="p6">Add</button>
</section>
`;

exports[`app mounts without errors 2`] = `
<section class="">
  <todo-list-stub todos="[object Object],[object Object]"></todo-list-stub>
  <hr><input class="p6" type="text"><button class="p6">Add</button>
</section>
`;
